core/src/main/java/org/apache/iceberg/actions/RewriteDataFilesCommitUtil.java

```diff
@@ -0,0 +1,208 @@
+
+  /**
+   * An async service which allows for committing multiple file groups as their rewrites complete. The service also
+   * allows for partial-progress since commits can fail. Once the service has been closed no new file groups should not
+   * be offered.
+   * @param rewritesPerCommit number of file groups to include in a commit
+   * @return the service for handling commits
+   */
+  public CommitService service(int rewritesPerCommit) {
+    return new CommitService(rewritesPerCommit);
+  }
+
+  public class CommitService implements Closeable {
+    private final ExecutorService committerService;
+    private final ConcurrentLinkedQueue<RewriteFileGroup> completedRewrites;
+    private final List<RewriteFileGroup> committedRewrites;
+    private final int rewritesPerCommit;
+    private AtomicBoolean running = new AtomicBoolean(false);
+
+    CommitService(int rewritesPerCommit) {
+      LOG.info("Creating commit service for table {} with {} groups per commit", table, rewritesPerCommit);
+      this.rewritesPerCommit = rewritesPerCommit;
+
+      committerService = Executors.newSingleThreadExecutor(new ThreadFactoryBuilder()
+          .setNameFormat("Committer-Service")
+          .build());
+
+      completedRewrites = Queues.newConcurrentLinkedQueue();
+      committedRewrites = Lists.newArrayList();
+    }
+
+    /**
+     * Starts a single threaded executor service for handling file group commits.
+     */
+    public void start() {
+      Preconditions.checkState(running.compareAndSet(false, true), "Rewrite Commit service already started");
+      LOG.info("Starting commit service for {}", table);
+      // Partial progress commit service
+      committerService.execute(() -> {
+        while (running.get() || completedRewrites.size() > 0) {
+          Thread.yield();
+          // Either we have a full commit group, or we have completed writing and need to commit what is left over
+          if (completedRewrites.size() >= rewritesPerCommit || (!running.get() && completedRewrites.size() > 0)) {
+
+            Set<RewriteFileGroup> batch = Sets.newHashSetWithExpectedSize(rewritesPerCommit);
+            for (int i = 0; i < rewritesPerCommit && !completedRewrites.isEmpty(); i++) {
+              batch.add(completedRewrites.poll());
+            }
+
+            try {
+              commitOrClean(batch);
+              committedRewrites.addAll(batch);
+            } catch (Exception e) {
+              LOG.error("Failure during rewrite commit process, partial progress enabled. Ignoring", e);
+            }
+          }
+        }
+      });
+    }
+
+    /**
+     * Places a file group in the queue to be asynchronously committed either when the queue has enough elements
+     * to do a batch of size {@link #rewritesPerCommit} or the service has been closed.
+     * @param group file group to eventually be committed
+     */
+    public void offer(RewriteFileGroup group) {
+      LOG.info("Offered to commit service: {}", group);
```

Comment:
I think this should be debug rather than info. We don't need to know when groups are handed off in normal operation, right?

Reply:
yep, That was from me debugging I forgot to lower the severity when i was done